### DISCUSSION TITLE
Better memory management when running sbt in a docker or a container

### DISCUSSION THREAD

**Discussion by seveneves** [sbt/sbt#6878] [Ideas]
JVM has been supporting integration with docker or a container environment like kubernetes for some time already. It can read  the cgroup limits and auto-assign corresponding memory usage. The amount of memory allocation can be tweaked via JVM options `MinRAMPercentage`, `MaxRAMPercentage`, `InitialRAMPercentage`.  Read more in this blog post [JVM in a Container](https://merikan.com/2019/04/jvm-in-a-container/)

However, these options are ignored if there are hard limits set on JVM via memory flags like `-Xmx`. And sbt does add them by default when starting a new JVM. 

It should be allowed to launch sbt with a flag `--container` for example, when the launcher script won't apply any limits to memory of JVM and allow it to read the cgroup limits directly.

**Comment by eed3si9n** [comment]
Sounds like a good idea. As a name `--nomem` might be more specific than `--container`.

If anyone wants to contribute this feature the the relevant parts are:
- https://github.com/sbt/sbt/blob/develop/sbt
- https://github.com/sbt/sbt/blob/develop/launcher-package/src/universal/bin/sbt.bat
- https://github.com/sbt/sbt/blob/f12a199a1aa58496ade139adcf8a7a8eba94b968/launcher-package/integration-test/src/test/scala/ScriptTest.scala#L85-L111

**Reply by seveneves** [reply]
While I was looking into implementing this, I found an interesting behavior of the `sbt` launcher.

```bash
addDefaultMemory() {
  # if we detect any of these settings in ${JAVA_OPTS} or ${JAVA_TOOL_OPTIONS} we need to NOT output our settings.
  # The reason is the Xms/Xmx, if they don't line up, cause errors.
  if [[ "${java_args[@]}" == *-Xmx* ]] || \
     [[ "${java_args[@]}" == *-Xms* ]] || \
     [[ "${java_args[@]}" == *-Xss* ]] || \
     [[ "${java_args[@]}" == *-XX:+UseCGroupMemoryLimitForHeap* ]] || \
     [[ "${java_args[@]}" == *-XX:MaxRAM* ]] || \
     [[ "${java_args[@]}" == *-XX:InitialRAMPercentage* ]] || \
     [[ "${java_args[@]}" == *-XX:MaxRAMPercentage* ]] || \
     [[ "${java_args[@]}" == *-XX:MinRAMPercentage* ]]; then
    :
  elif [[ "${JAVA_TOOL_OPTIONS}" == *-Xmx* ]] || \
       [[ "${JAVA_TOOL_OPTIONS}" == *-Xms* ]] || \
       [[ "${JAVA_TOOL_OPTIONS}" == *-Xss* ]] || \
       [[ "${JAVA_TOOL_OPTIONS}" == *-XX:+UseCGroupMemoryLimitForHeap* ]] || \
       [[ "${JAVA_TOOL_OPTIONS}" == *-XX:MaxRAM* ]] || \
       [[ "${JAVA_TOOL_OPTIONS}" == *-XX:InitialRAMPercentage* ]] || \
       [[ "${JAVA_TOOL_OPTIONS}" == *-XX:MaxRAMPercentage* ]] || \
       [[ "${JAVA_TOOL_OPTIONS}" == *-XX:MinRAMPercentage* ]] ; then
    :
  elif [[ "${sbt_options[@]}" == *-Xmx* ]] || \
       [[ "${sbt_options[@]}" == *-Xms* ]] || \
       [[ "${sbt_options[@]}" == *-Xss* ]] || \
       [[ "${sbt_options[@]}" == *-XX:+UseCGroupMemoryLimitForHeap* ]] || \
       [[ "${sbt_options[@]}" == *-XX:MaxRAM* ]] || \
       [[ "${sbt_options[@]}" == *-XX:InitialRAMPercentage* ]] || \
       [[ "${sbt_options[@]}" == *-XX:MaxRAMPercentage* ]] || \
       [[ "${sbt_options[@]}" == *-XX:MinRAMPercentage* ]] ; then
    :
  else
    addMemory $sbt_default_mem
  fi
}
```

It will not add any defaults if there is any of the above mentioned flags specified. Which means, that I can call 

```bash 
 JAVA_TOOL_OPTIONS="-XX:MaxRAMPercentage=90" sbt
```
To have the desired effect of `--nomem` flag. 

Flag `--nomem` would be great to have for more explicit behavior instead of relying on "internal" knowledge of how it is implemented.
